### PR TITLE
drivers: i2c: esp32: fix timing and stop bit

### LINKED ISSUE #34015: cfb sample "Device not found" for esp32 when SSD1306 is enabled
**Describe the bug**
When building a project with the SSD1306 enabled and set up, I get this output:

**To Reproduce**
1 - add `CONFIG_SSD1306=y` to prj.conf
2- add `arduino headers` and `arduino_I2C` to `esp32.dts`
```
`{
	 arduino_header: connector {
		 compatible = "arduino-header-r3";
		 #gpio-cells = <2>;
		 gpio-map-mask = <0xffffffff 0xffffffc0>;
		 gpio-map-pass-thru = <0 0x3f>;
		 gpio-map = <0 0 &gpio0 4 0>,	/* A0 */
				<1 0 &gpio0 0 0>,	/* A1 */
				<2 0 &gpio0 2 0>,	/* A2 */
				<3 0 &gpio0 15 0>,	/* A3 */
				///////////I2C///////////
				<4 0 &gpio0 13 0>,	/* A4 */
				<5 0 &gpio0 12 0>,	/* A5 */
				<6 0 &gpio0 1 0>,	/* D0 */
				<7 0 &gpio0 3 0>,	/* D1 */
				///////////////SPI/////////////////
				<8 0 &gpio0 18 0>,	/* D10 */
				<9 0 &gpio0 27 0>,	/* D11 */
				<10 0 &gpio0 19 0>,	/* D12 */
				<11 0 &gpio0 5 0>;	/* D13 */
				/////////////////////////////////			   
	 };
 };
 
 arduino_i2c: &i2c0 {
	 status = "okay";
	 clock-frequency = <I2C_BITRATE_FAST>;
	 sda-pin = <4>;
	 scl-pin = <15>;
 };
 ////////////////////////////////////////////////////////`
```
2-Build using west build -b esp32 samples\display\cfb -- -DSHIELD=ssd1306_128x64

**Expected behavior**
display a message on oled display 

**output**
```
`
▒*** Booting Zephyr OS version 2.5.99  ***source...
Device not found
[00:00:00.437,000] <err> ssd1306: Failed to initialize device!

`
```

**Environment (please complete the following information):**
 OS: Windows
Toolchain Zephyr SDK

### PR DESCRIPTION
**Pull request by sylvioalves**
Current implementation fails due to missing stop bit
when reading data. This fixes it and refactor the implementation
by adding k_busy_wait when waiting I2C bus completion.

Also, this implementation is based on esp-idf v3.0.
It will be refactored based on latest esp-idf v4.3 using proper
low level calls.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

Fixes #34015

### CODE REVIEW

**Comment by sylvioalves** [comment]
I tagged this to v2.6.0 if still within the window frame.

File: drivers/i2c/i2c_esp32.c [hunk 1]
```diff
@@ -276,32 +273,14 @@ static inline void i2c_esp32_reset_fifo(const struct i2c_esp32_config *config)
...
-	/* Start transmission and wait for the ISR to give the semaphore */
-	sys_set_bit(I2C_CTR_REG(config->index), I2C_TRANS_START_S);
-	if (k_sem_take(&data->fifo_sem, K_MSEC(I2C_ESP32_TIMEOUT_MS)) < 0) {
-		return -ETIMEDOUT;
+	if (*counter == 0) {
+		return 1;
 	}
 
+	(*counter)--;
+	k_busy_wait(1);
```
Comment (mniestroj) [inline drivers/i2c/i2c_esp32.c:283]: Why is it busy wait instead of k_sleep?

Comment (sylvioalves) [reply drivers/i2c/i2c_esp32.c:283]: Using `k_sleep(K_USEC(1))` increases latency from 10us to 100us between byte read call.

Comment (mniestroj) [reply drivers/i2c/i2c_esp32.c:283]: So where does 1ms (`#define I2C_ESP32_TIMEOUT_USEC  1000`) timeout come from?

>This timeout is not for the whole transfer, only to wait I2C channel to become idle.

Become idle from which state?

>Using k_sleep(K_USEC(1)) increases latency from 10us ....

10us, not 1us? Won't that 1ms timeout become 10ms then?

Comment (sylvioalves) [reply drivers/i2c/i2c_esp32.c:283]: @mniestroj , this 1ms wait time comes from esp-idf 3.0 I2C implementation, where it checks I2C bus line errors (or busy).  
Regarding the `k_sleep`, it works based on system tick, which would not be small as ~1us (which we do have with nop operations or `k_busy_wait()`).

**Comment by cfriedt** [comment]
@sylvioalves - I re-requested review from @mniestroj . I'll approve / merge if he is able to re-review.

**Review by mniestroj — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**